frontend/ui/widget/filechooser.lua

```diff
@@ -237,81 +355,35 @@ function FileChooser:getSortingFunction(collate, reverse_collate)
+    local collate = self:getCollate()
     local dirs, files = self:getList(path, collate)
     return self:genItemTable(dirs, files, path)
 end
 
-function FileChooser.isCollateNotForMixed(collate)
-    return collate == "size" or collate == "type"
-        or collate == "percent_unopened_first" or collate == "percent_unopened_last"
-end
-
 function FileChooser:genItemTable(dirs, files, path)
-    local collate = G_reader_settings:readSetting("collate")
+    local collate = self:getCollate()
     local collate_mixed = G_reader_settings:isTrue("collate_mixed")
     local reverse_collate = G_reader_settings:isTrue("reverse_collate")
     local sorting = self:getSortingFunction(collate, reverse_collate)
-    local collate_not_for_mixed = self.isCollateNotForMixed(collate)
-    if collate_not_for_mixed or not collate_mixed then
-        table.sort(files, sorting)
-        if collate_not_for_mixed then -- keep folders sorted by name not reversed
-            sorting = self:getSortingFunction("strcoll")
-        end
-        table.sort(dirs, sorting)
-    end
 
     local item_table = {}
-
-    for i, dir in ipairs(dirs) do
-        local text, bidi_wrap_func, mandatory
-        if dir.text == "./." then -- added as content of an unreadable directory
-            text = _("Current folder not readable. Some content may not be shown.")
-        else
-            text = dir.text.."/"
-            bidi_wrap_func = BD.directory
-            if path then -- file browser or PathChooser
-                mandatory = self:getMenuItemMandatory(dir)
-            end
-        end
-        table.insert(item_table, {
-            text = text,
-            attr = dir.attr,
-            bidi_wrap_func = bidi_wrap_func,
-            mandatory = mandatory,
-            path = dir.fullpath,
-        })
-    end
-
-    -- set to false to show all files in regular font
-    -- set to "opened" to show opened files in bold
-    -- otherwise, show new files in bold
-    local show_file_in_bold = G_reader_settings:readSetting("show_file_in_bold")
-
-    for i, file in ipairs(files) do
-        local file_item = {
-            text = file.text,
-            attr = file.attr,
-            bidi_wrap_func = BD.filename,
-            mandatory = self:getMenuItemMandatory(file, collate),
-            path = file.fullpath,
-            is_file = true,
-        }
-        if show_file_in_bold ~= false then
-            file_item.bold = file.opened
-            if show_file_in_bold ~= "opened" then
-                file_item.bold = not file_item.bold
-            end
-        end
-        if self.filemanager and self.filemanager.selected_files and self.filemanager.selected_files[file.fullpath] then
-            file_item.dim = true
-        end
-        table.insert(item_table, file_item)
-    end
-
-    if not collate_not_for_mixed and collate_mixed then
+    if collate.can_collate_mixed and collate_mixed then
+        table.move(dirs, 1, #dirs, 1, item_table)
```

Comment:
The first usage in the project!

Reply:
I was really hoping LuaJIT would do some optimized `memcpy`, but that doesn't appear to be the case. Still, it should at least be faster than writing the loop ourselves, and less lines of code as well.

https://github.com/LuaJIT/LuaJIT/blob/43d0a19158ceabaa51b0462c1ebc97612b420a2e/src/lib_table.c#L132-L150